services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaServiceITest.kt

```diff
@@ -210,6 +207,97 @@ class HankeKayttajaServiceITest : DatabaseTest() {
+
+        assertThat(kayttajaTunnisteRepository.findAll()).isEmpty()
+        assertThat(hankeKayttajaRepository.findAll()).isEmpty()
+    }
+
+    @Test
+    fun `saveNewTokensFromHanke creates tokens for unique ones`() {
+        val hanke =
+            hankeFactory.save(
+                HankeFactory.create()
+                    .withYhteystiedot(
+                        // each has a duplicate
+                        omistajat = listOf(1, 1),
+                        rakennuttajat = listOf(2, 2),
+                        toteuttajat = listOf(3, 3),
+                        muut = listOf(4, 4)
+                    )
+            )
+        assertThat(hanke.extractYhteystiedot()).hasSize(8)
+
+        hankeKayttajaService.saveNewTokensFromHanke(hanke)
+
+        val tunnisteet: List<KayttajaTunnisteEntity> = kayttajaTunnisteRepository.findAll()
+        val kayttajat: List<HankeKayttajaEntity> = hankeKayttajaRepository.findAll()
+        assertThat(tunnisteet).hasSize(4) // 4 yhteyshenkilo subcontacts.
+        assertThat(kayttajat).hasSize(4)
+        assertTunnisteet(tunnisteet)
+        assertKayttajat(kayttajat, hanke.id)
+    }
+
+    @Test
+    fun `saveNewTokensFromHanke with pre-existing permissions does not create duplicate`() {
+        val hanke = hankeFactory.save(HankeFactory.create())
+        saveUserAndPermission(hanke, "Existing User One", "ali.kontakti@meili.com")
+
+        hankeKayttajaService.saveNewTokensFromHanke(
+            hanke.apply { this.omistajat.add(HankeYhteystietoFactory.create()) }
+        )
+
+        val tunnisteet = kayttajaTunnisteRepository.findAll()
+        assertThat(tunnisteet).isEmpty()
+        assertTunnisteet(tunnisteet)
+        val kayttajat = hankeKayttajaRepository.findAll()
+        assertThat(kayttajat).hasSize(1)
+        assertThat(kayttajat.map { it.sahkoposti })
+            .containsExactly(
+                "ali.kontakti@meili.com",
+            )
+    }
+
+    private fun assertTunnisteet(tunnisteet: List<KayttajaTunnisteEntity>) =
+        assertThat(tunnisteet).each { t ->
+            t.transform { it.id }.isNotNull()
+            t.transform { it.role }.isEqualTo(Role.KATSELUOIKEUS)
+            t.transform { it.createdAt }.isRecent()
+            t.transform { it.sentAt }.isNull()
+            t.transform { it.tunniste }.matches(Regex(kayttajaTunnistePattern))
+            t.transform { it.hankeKayttaja }.isNotNull()
+        }
+
+    private fun assertKayttajat(kayttajat: List<HankeKayttajaEntity>, hankeId: Int?) =
+        assertThat(kayttajat).each { k ->
+            k.transform { it.id }.isNotNull()
+            k.transform { it.nimi }.isIn(*expectedNames)
+            k.transform { it.sahkoposti }.isIn(*expectedEmails)
+            k.transform { it.hankeId }.isEqualTo(hankeId)
+            k.transform { it.permission }.isNull()
+            k.transform { it.kayttajaTunniste }.isNotNull()
+        }
```

Comment:
Using custom asserts gives better error messages if one of the asserts fails.
```suggestion
    private fun Assert<List<KayttajaTunnisteEntity>>.areValidTunnisteet() =
        this.each { t ->
            t.transform { it.id }.isNotNull()
            t.transform { it.role }.isEqualTo(Role.KATSELUOIKEUS)
            t.transform { it.createdAt }.isRecent()
            t.transform { it.sentAt }.isNull()
            t.transform { it.tunniste }.matches(Regex(kayttajaTunnistePattern))
            t.transform { it.hankeKayttaja }.isNotNull()
        }

    private fun Assert<List<HankeKayttajaEntity>>.areValidKayttajat(hankeId: Int?) =
        this.each { k ->
            k.transform { it.id }.isNotNull()
            k.transform { it.nimi }.isIn(*expectedNames)
            k.transform { it.sahkoposti }.isIn(*expectedEmails)
            k.transform { it.hankeId }.isEqualTo(hankeId)
            k.transform { it.permission }.isNull()
            k.transform { it.kayttajaTunniste }.isNotNull()
        }
```

Use like
```
        assertThat(tunnisteet).areValidTunnisteet()
        assertThat(kayttajat).areValidKayttajat(hanke.id)
```

Reply:
I'm not sure about my naming, though.

Reply:
Changed as suggested. I chose function name is "areValid()". tunnisteet).areValidTunnisteet() felt repetitive.